src/base/dof_map_constraints.C

```diff
@@ -431,6 +431,9 @@ private:
         std::vector<bool> is_boundary_node(elem->n_nodes(), false),
           is_boundary_edge(elem->n_edges(), false),
           is_boundary_side(elem->n_sides(), false);
+
+        // We also maintain a separate list of nodeset-based boundary nodes
+        std::vector<bool> is_boundary_nodeset(elem->n_nodes(), false);
```

Comment:
I don't think any locking is needed.  Basically the way that const ```vector<bool>::operator[]``` works is to bitwise-and the appropriate char against a mask.  You can do that as many times as you want concurrently and not break anything.

With non-const ```vector<bool>::operator[]```, by contrast, under the hood ```=``` is similar to ```++``` for integers: it has to first read data (that char), then modify that data (with a bitwise-or for true or bitwise-and for false), then write the data back.  If someone else tries to do the same thing at the same time, then you risk them both getting the same initial data and then only one of the two modifications gets written back.